src/libxively/event_dispatcher/xi_time_event.c

```diff
@@ -0,0 +1,426 @@
+ * an element in both directions in the vector. It is used by
+ * xi_time_event_bubble_and_sort_down and xi_time_event_bubble_and_sort_up.
+ *
+ * @param vector - input vector, container of time events
+ * @param lhs - value that will appear on the left hand side of the stop condition
+ * @param rhs - value that will appear on the right hand side of the stop condition
+ * @param elem_index_var_pos - points which: lhs or rhs represents the index of an element
+ * @param dir - direction of bubbling - +1 stands for up -1 means down
+ * @return new index of bubbled element
+ */
+static xi_vector_index_type_t xi_time_event_bubble_core( xi_vector_t* vector,
+                                                         xi_vector_index_type_t lhs,
+                                                         xi_vector_index_type_t rhs,
+                                                         uint8_t elem_index_var_pos,
+                                                         int8_t dir )
+{
+    xi_vector_index_type_t* indexes[] = {&lhs, &rhs};
+
+    while ( lhs < rhs )
+    {
+        const xi_vector_index_type_t elem_index = *indexes[elem_index_var_pos];
+
+        const xi_time_event_t* lhs_element =
+            ( xi_time_event_t* )vector->array[dir < 0 ? elem_index : elem_index + dir]
+                .selector_t.ptr_value;
+        const xi_time_event_t* rhs_element =
+            ( xi_time_event_t* )vector->array[dir < 0 ? elem_index + dir : elem_index]
+                .selector_t.ptr_value;
+
+        if ( lhs_element->time_of_execution < rhs_element->time_of_execution )
+        {
+            /* if the elements are not in the right order lets swap them */
+            xi_swap_time_events( vector, lhs_element->position, rhs_element->position );
+
+            /* update the indexes */
+            *indexes[elem_index_var_pos] += dir;
+        }
+        else
+        {
+            /* thanks to the container invariant on the order of the elemements we can
+             * break at this moment */
+            break;
+        }
+    }
+
+    return *indexes[elem_index_var_pos];
+}
+
+/**
+ * @brief xi_time_event_bubble_and_sort_down
+ *
+ * Helper function used for insertion and re-insertion ( cancel opertaion ). It uses the
+ * bubble sort algorithm to put the element on the proper position.
+ *
+ * @note: invariant of this container - it's elements are sorted using the time of
+ * execution so:
+ * vector[i].time_of_execution <= vector[i+1].time_of_execution
+ *
+ * @param vector
+ * @param index
+ * @return new index of the element after bubbling it
+ */
+static xi_vector_index_type_t
+xi_time_event_bubble_and_sort_down( xi_vector_t* vector, xi_vector_index_type_t index )
+{
+    /* PRE-CONDITIONS */
+    assert( NULL != vector );
+    assert( index >= 0 );
+    assert( index < vector->elem_no );
+
+    return xi_time_event_bubble_core( vector, 0, index, 1, -1 );
```

Comment:
do I understand this well that this brings (index)th element to the 0th position?